### DISCUSSION TITLE
Task 7: Retrieve historical price data from on chain

### DISCUSSION THREAD

**Discussion by icmoore** [defipy-devs/defipy#13] [Ideas]
While the 0x API can read live data, it does not have the ability to read historical data. Not sure quite how this can be achieved, whether it can be read off the chain directly, or via an API service. 

Ideas and suggestions are welcome for this task

**Comment by ABusyHippie** [comment]
DeFiLlama's API is quite comprehensive and allows grabbing historical data and prices by pair, so we can also add historical/live prices of token pairs not just in stablecoin terms (like the 0x API limitation). It's also free below a certain rate limit and has lots of endpoints and data that could be helpful

https://defillama.com/docs/api

**Comment by ABusyHippie** [comment]
I will start working on this in my own fork [here](https://github.com/ABusyHippie/uniswappy/tree/ethdenver)

**Comment by ABusyHippie** [comment]
@icmoore does web3scout meet this requirement? Should we close?

**Reply by ABusyHippie** [reply]
Still needs tailoring but meets this requirement generally, closing task
